lib/fizzy/execute.cpp

```diff
@@ -301,7 +301,8 @@ inline DstT extend(SrcT in) noexcept
 template <typename DstT, typename SrcT = DstT>
 inline bool load_from_memory(bytes_view memory, Stack<uint64_t>& stack, const uint8_t*& immediates)
 {
-    const auto address = static_cast<uint32_t>(stack.pop());
+    // Addressing is 32-bit, but we keep the value as 64-bit to detect overflows.
+    const auto address = stack.pop() & 0xffffffff;
```

Comment:
I don't think the mask is needed at all. If I'm wrong, can you add a test demonstrating this case?

Reply:
If the point of the change is to change the type of `address`, maybe makes sense to state the type explicitly instead of `auto`?

Reply:
Or maybe it should be instead solved with a cast at the point of comparison with `memory.size()`

Reply:
> If the point of the change is to change the type of address, maybe makes sense to state the type explicitly instead of auto?

The calculation should be done on `uin64_t` to detect overflows. Both `address` and `offset` are limited to 32-bits.

Reply:
I think I would do
```
 if ((static_cast<uint64_t>(address) + offset + sizeof(SrcT)) > memory.size())
```

Reply:
> I don't think the mask is needed at all. If I'm wrong, can you add a test demonstrating this case?

I don't remember our stance on the stack. Did we say that we perform no validation (except casting) and rely on the validation algorithm to ensure the stack is sound (e.g. always the same type or with explicit conversion opcodes)?

Reply:
Yes, my understanding is that the validation should ensure that the stack has always correct type for the operation.

In particular, the spec of `load` execution states:

![image](https://user-images.githubusercontent.com/1863135/75017538-819ea080-548d-11ea-9502-a0df2765c0b5.png)

Reply:
In current version it is assumed the types match and moreover `i32` are zero extended to 64 bits when placed on the stack.

There is an alternative design we can try later.

I recommend not to go defensive here (and every other place) to make sure it works as designed. You can always add an assert though.

However, there is a catch. It might be ok to produce wrong results when executing invalid wasm. But might not be ok for the interpreter to crash in this case.